features/steps/steps.py

```diff
@@ -331,16 +406,25 @@ def step_impl(context, relationship_type, entity):
 
 @given('A file with {field} "{values}"')
 def step_impl(context, field, values):
-    values = list(map(str.lower, map(lambda s: s.strip('"'), values.split(' or '))))
+    values = strip_split(values, strp = '"', splt = ' or ')
     if field == "Model View Definition":
         conditional_lowercase = lambda s: s.lower() if s else None
         applicable = conditional_lowercase(get_mvd(context.model)) in values
     elif field == "Schema Identifier":
         applicable = context.model.schema.lower() in values
     else:
         raise NotImplementedError(f'A file with "{field}" is not implemented')
-
+ 
     context.applicable = getattr(context, 'applicable', True) and applicable
+    
+@given('Its values')
+def step_impl(context):
+    context._push()
+    instances_unpacked = unpack_sequence_of_entities(
```

Comment:
Instead of this `unpack_sequence_of_entities`, I'd actually expect we'd be using this `map_state()` function. Did you try? What were the issues?

Reply:
Yes, I had a bit of trouble figuring it out. For example take the following two instances, both packed in a tuple
`
[(#22=IfcRigidOperation(#11,#21,IfcLengthMeasure(35010.),IfcLengthMeasure(1560.),$),), (#24=IfcRigidOperation(#23,#21,IfcLengthMeasure(23684.),IfcLengthMeasure(1564.),$),)]`

With the desired output: 
`
[#22=IfcRigidOperation(#11,#21,IfcLengthMeasure(35010.),IfcLengthMeasure(1560.),$), #24=IfcRigidOperation(#23,#21,IfcLengthMeasure(23684.),IfcLengthMeasure(1564.),$)]`


I tried it with
`instances_unpacked =list(map_state(context.instances, lambda i: i)) # or even without list(...)`
which returns (the same as context.instances)
`[(#22=IfcRigidOperation(#11,#21,IfcLengthMeasure(35010.),IfcLengthMeasure(1560.),$),), (#24=IfcRigidOperation(#23,#21,IfcLengthMeasure(23684.),IfcLengthMeasure(1564.),$),)]`

However, when trying
`instances_unpacked = list(map_state(context.instances, lambda i: i[0]))`
it returns
`[(#11=IfcGeometricRepresentationContext($,'Model',3,1.E-05,#9,#10),), (#23=IfcGeometricRepresentationContext($,'Model',3,1.E-05,#9,#10),)]`
Which suggest that it gets to the ifcopenshell.entity_instance, of which i[0] refers to the first item in that instance
I tried some more things, but I could not seem to unpack it proper this way. And with the unpack_sequence_of_entities function it worked. Do I miss something?

Reply:
I'm not convinced there is an issue with that.

Yes, `HasCoordinateOperation` is a `set[0:1]` so in the step1 below is where the additional "list-wrapping" is introduced.

```
[IfcGeometricRepresentationContext_0, IfcGeometricRepresentationContext_1, IfcGeometricRepresentationContext_2]
[[IfcCoordinateOperation_0], [IfcCoordinateOperation_1], []]
[[{TargetCRS:IfcCoordinateReferenceSystem_0, Eastings: 1, ...}], [{TargetCRS:IfcCoordinateReferenceSystem_1, Eastings:2 , ...}]]
```

But the constraint on identical values is still the same according to the python semantics. Trivial example:

```
>>> 1 == 1
True
>>> [1] == [1]
True
```

So even if the values are wrapped in a list of length one (in the bigger list of all instance value), we'd still expect the same result:

```
>>> set([(1,),(1,)])
{(1,)}
>>> set([(1,),(2,)])
{(1,), (2,)}
```

But in the case above I had to use tuples, because lists are mutable in python and therefore not hashable and therefore not able to be converted to set.

```
>>> set([[1],[2]])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'list'
```

But you can still do:

```
>>> vs = [[1], [1], [1]]
>>> all(v == vs[0] for v in vs[1:])
True
```

> With the desired output:

So this I'm basically trying to question. The values wrapped in a tuple or list doesn't change in this case the ability to test them to be identical.

Reply:
I understand, thanks for the explanation. 
Even is the output is not desired in this case, I still don't fully understand why map_state returns a tuple. 
But I'll wrap my head around that outside this PR. 

```
The unpack_sequence_of_entities function is replaced by map_state. 
    context.instances = map_state(context.instances, lambda inst: do_try(
        lambda: inst.get_info(recursive=True, include_identifier=False, ignore=excluding), None))
# as you suggested, the output is [(A),(B)] ==  [(A),(B)]  == True 
```
https://github.com/buildingSMART/ifc-gherkin-rules/pull/35/commits/498a5ca9b9acc0190784ba7a76d423e54dc30de8